### ISSUE TITLE
 fields with `this.hasOne` are not insertedOrUpdated correctly and are given a value from a different relationship

### ISSUE THREAD

**Issue by JasonLandbridge** [vuex-orm/vuex-orm#437]
TL;DR; fields with `this.hasOne` are not insertedOrUpdated correctly and are given a value from a different relationship. 

I have a Resume model with the following fields:

```
    ....
      qualities: this.hasMany(DualInputValue, "resume_id"),
      thriveProfessionally: this.hasOne(DualInputValue, "resume_id"),
      whatColleguesNeedToKnow: this.hasOne(DualInputValue, "resume_id"),
      thingsInFutureProjects: this.hasMany(DualInputValue, "resume_id"),
```

All of these use the same model DualInputValue, this model has the following fields:

```
      id: this.increment(),
      resume_id: this.attr(null),
      type: this.attr(""),
      dutch_value: this.attr(""),
      english_value: this.attr(""),
```

Basically, the resume contains two input fields where the user can give a Dutch and English translation for the same form entry. 

Now here is the problem, when I create a new Resume I give several default values only they end up mixed. This only goes for the Resume fields which have a "this.hasOne(DualInputValue, "resume_id")" value:

How I create the new Resume:
``` js
// mutations
const mutations = {
 createResume(state: any, resume_id: number) {
    Resume.insertOrUpdate({
      data: {
        id: resume_id,
        qualities: [{ resume_id, type: "quality" }],
        thriveProfessionally: {
           resume_id,
           type: "thriveProfessionally",
           dutch_value: "LOL",
           english_value: "BBQ"
        },
         whatColleguesNeedToKnow: {
           resume_id,
           type: "whatColleguesNeedToKnow",
          dutch_value: "LOL",
           english_value: "BBQ"
        },
        thingsInFutureProjects: [
          {
            resume_id,
            type: "thingsInFutureProjects"
          },
          {
            resume_id,
            type: "thingsInFutureProjects"
          },
          {
            resume_id,
            type: "thingsInFutureProjects"
          }
        ],
        certificates: [
          {
            resume_id,
            type: "certificates"
          }
        ],
        competenties: [
          {
            resume_id,
            type: "competenties"
          }
        ],
        studies: [
          {
            resume_id,
            type: "studies"
          }
        ],
        hobbiesAndInterests: [
          {
            resume_id,
            type: "hobbiesAndInterests"
          }
        ],
      }
    });
  }
}
```

And then I would like an output of all data including the relationships of this resume:

```js
// getters
const getters = {
  resumeData: (state: any) => (id: number) => {
    return Resume.query()
      .whereId(id)
      .withAll()
      .first();   
    }
};
```

But I end up with the following results:

```js
{
  "id": 431,
  "firstName": "",
  "lastName": "",
  "region": "",
  "nationality": "",
  "driversLicenses": [],
  "languages": [],
  "educations": [],
  "qualities": [],
  "thriveProfessionally": {
    "id": 18,
    "resume_id": 431,
    "type": "hobbiesAndInterests",
    "dutch_value": "",
    "english_value": ""
  },
  "whatColleguesNeedToKnow": {
    "id": 18,
    "resume_id": 431,
    "type": "hobbiesAndInterests",
    "dutch_value": "",
    "english_value": ""
  },
  "thingsInFutureProjects": [
    {
      "id": 12,
      "resume_id": 431,
      "type": "thingsInFutureProjects",
      "dutch_value": "",
      "english_value": ""
    },
    {
      "id": 13,
      "resume_id": 431,
      "type": "thingsInFutureProjects",
      "dutch_value": "",
      "english_value": ""
    },
    {
      "id": 14,
      "resume_id": 431,
      "type": "thingsInFutureProjects",
      "dutch_value": "",
      "english_value": ""
    }
  ],
  "certificates": [
    {
      "id": 15,
      "resume_id": 431,
      "type": "certificates",
      "dutch_value": "",
      "english_value": ""
    }
  ],
  "competenties": [
    {
      "id": 16,
      "resume_id": 431,
      "type": "competenties",
      "dutch_value": "",
      "english_value": ""
    }
  ],
  "studies": [
    {
      "id": 17,
      "resume_id": 431,
      "type": "studies",
      "dutch_value": "",
      "english_value": ""
    }
  ],
  "hobbiesAndInterests": [
    {
      "id": 18,
      "resume_id": 431,
      "type": "hobbiesAndInterests",
      "dutch_value": "",
      "english_value": ""
    }
  ]
}
```

What happens is that  "thriveProfessionally" and "whatColleguesNeedToKnow" are both given a type of "hobbiesAndInterests". Weirdly enough, if I change the order of the data in the createResume function, then a different wrong type is inserted. 

It seems to take the last field given in the data {}, takes that type and then uses that for fields which have a "hasOne" relationship. 

Thank you for reading!

**Comment by kiaking** [comment]
#437 Hi! thank you for the report.

Well, this behavior kind of happening by nature of the RDM data structure. The problem is that you have 2 `hasOne` relationship.

Has One relationship assume that Resume only has one DualInputValue model relationship in the store. With your example, there will be many DualInputValue with same `resume_id`. And Has One relationship will stop looking up DualInputVal when it finds the first matching record. That is why you will get a different result if you change the array order.

In this case, you shouldn't be using Has One relationship, because the relationship is not only one, it has many. Instead, filter the result based on `type` filed. You could query relationships like this.

```js
Resume.query().with('thriveProfessionally', (query) => {
  query.where('type', 'thriveProfessionally')
}).with('whatColleguesNeedToKnow', (query) => {
  query.where('type', 'whatColleguesNeedToKnow')
}).get()
```

**Comment by JasonLandbridge** [comment]
Thank you so much for explaining this! I managed to solve this thanks to your explanation!
